File: untitled folder/untitled folder 2/flutter_secure_biometrics/android/src/main/kotlin/com/example/flutter_secure_biometrics/LivenessDetector.kt
package com.example.flutter_secure_biometrics

import android.content.Context
import android.hardware.camera2.CameraAccessException
import android.hardware.camera2.CameraManager
import android.os.Handler
import android.os.Looper
import kotlin.random.Random

class LivenessDetector {
    
    private val handler = Handler(Looper.getMainLooper())
    
    fun performDetection(context: Context, biometricType: String): Boolean {
        return when (biometricType) {
            "face" -> performFaceLivenessDetection(context)
            "fingerprint" -> performFingerprintLivenessDetection(context)
            "voice" -> performVoiceLivenessDetection(context)
            "heartRate" -> performHeartRateLivenessDetection(context)
            else -> true // Default to pass for unsupported types
        }
    }
    
    private fun performFaceLivenessDetection(context: Context): Boolean {
        return try {
            val cameraManager = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
            val frontCameraId = getFrontCameraId(cameraManager)
            
            if (frontCameraId != null) {
                // Simulate blink detection and 3D depth analysis
                // In real implementation, this would:
                // 1. Capture multiple frames
                // 2. Detect eye blinks
                // 3. Analyze facial movement
                // 4. Check for 3D depth using structured light or stereo vision
                simulateLivenessCheck()
            } else {
                false
            }
        } catch (e: CameraAccessException) {
            false
        }
    }
    
    private fun performFingerprintLivenessDetection(context: Context): Boolean {
        // Simulate pulse detection through fingerprint sensor
        // In real implementation, this would:
        // 1. Analyze blood flow patterns
        // 2. Detect pulse rhythm
        // 3. Check for temperature variations
        // 4. Analyze capacitive changes over time
        return simulatePulseDetection()
    }
    
    private fun performVoiceLivenessDetection(context: Context): Boolean {
        // Simulate voice challenge-response
        // In real implementation, this would:
        // 1. Generate random phrase for user to speak
        // 2. Analyze voice characteristics
        // 3. Check for natural speech patterns
        // 4. Detect breathing patterns in speech
        return simulateVoiceChallenge()
    }
    
    private fun performHeartRateLivenessDetection(context: Context): Boolean {
        // Simulate heart rate validation through camera
        // In real implementation, this would:
        // 1. Use camera and flash to detect pulse
        // 2. Analyze heart rate variability
        // 3. Check for consistent pulse patterns
        // 4. Validate against normal heart rate ranges
        return simulateHeartRateValidation()
    }
    
    private fun getFrontCameraId(cameraManager: CameraManager): String? {
        return try {
            cameraManager.cameraIdList.find { cameraId ->
                val characteristics = cameraManager.getCameraCharacteristics(cameraId)
                val facing = characteristics.get(android.hardware.camera2.CameraCharacteristics.LENS_FACING)
                facing == android.hardware.camera2.CameraCharacteristics.LENS_FACING_FRONT
            }
        } catch (e: Exception) {
            null
        }
    }
    
    private fun simulateLivenessCheck(): Boolean {
        // Simulate realistic liveness detection with 95% accuracy
        return Random.nextDouble() < 0.95
    }
    
    private fun simulatePulseDetection(): Boolean {
        // Simulate pulse detection with 90% accuracy
        return Random.nextDouble() < 0.90
    }
    
    private fun simulateVoiceChallenge(): Boolean {
        // Simulate voice challenge with 92% accuracy
        return Random.nextDouble() < 0.92
    }
    
    private fun simulateHeartRateValidation(): Boolean {
        // Simulate heart rate validation with 88% accuracy
        return Random.nextDouble() < 0.88
    }
    
    fun generateVoiceChallenge(): String {
        val challenges = listOf(
            "Please say: The quick brown fox jumps over the lazy dog",
            "Please say: Security is our top priority",
            "Please say: Today is a beautiful day for authentication",
            "Please say: My voice is my unique identifier",
            "Please say: Biometric security protects my data"
        )
        return challenges.random()
    }
    
    fun validateBlinkPattern(blinkTimings: List<Long>): Boolean {
        // Validate natural blink patterns
        // Normal blink rate: 15-20 blinks per minute
        // Blink duration: 100-400ms
        if (blinkTimings.size < 2) return false
        
        val intervals = blinkTimings.zipWithNext { a, b -> b - a }
        val averageInterval = intervals.average()
        
        // Check if blink rate is within normal range (2-6 seconds between blinks)
        return averageInterval in 2000.0..6000.0
    }
    
    fun analyzeFacialMovement(movementData: List<Pair<Float, Float>>): Boolean {
        // Analyze natural facial movements
        if (movementData.size < 5) return false
        
        val movements = movementData.zipWithNext { a, b ->
            val dx = b.first - a.first
            val dy = b.second - a.second
            kotlin.math.sqrt((dx * dx + dy * dy).toDouble())
        }
        
        val totalMovement = movements.sum()
        val averageMovement = movements.average()
        
        // Check for natural micro-movements (not too static, not too erratic)
        return totalMovement > 5.0 && averageMovement < 50.0
    }
}

File: untitled folder/untitled folder 2/flutter_secure_biometrics/ios/Classes/LivenessDetector.swift
import Foundation
import AVFoundation
import Vision
import CoreImage

class LivenessDetector: NSObject {
    
    private var captureSession: AVCaptureSession?
    private var videoOutput: AVCaptureVideoDataOutput?
    private let videoQueue = DispatchQueue(label: "videoQueue")
    
    func performDetection(type: String, completion: @escaping (Bool, Error?) -> Void) {
        switch type {
        case "face":
            performFaceLivenessDetection(completion: completion)
        case "fingerprint":
            performFingerprintLivenessDetection(completion: completion)
        case "voice":
            performVoiceLivenessDetection(completion: completion)
        case "heartRate":
            performHeartRateLivenessDetection(completion: completion)
        default:
            completion(true, nil) // Default to pass for unsupported types
        }
    }
    
    private func performFaceLivenessDetection(completion: @escaping (Bool, Error?) -> Void) {
        guard let frontCamera = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .front) else {
            completion(false, LivenessError.cameraNotAvailable)
            return
        }
        
        do {
            let captureSession = AVCaptureSession()
            let input = try AVCaptureDeviceInput(device: frontCamera)
            
            if captureSession.canAddInput(input) {
                captureSession.addInput(input)
            }
            
            let videoOutput = AVCaptureVideoDataOutput()
            videoOutput.setSampleBufferDelegate(self, queue: videoQueue)
            
            if captureSession.canAddOutput(videoOutput) {
                captureSession.addOutput(videoOutput)
            }
            
            self.captureSession = captureSession
            self.videoOutput = videoOutput
            
            // Start capture session
            DispatchQueue.global(qos: .userInitiated).async {
                captureSession.startRunning()
                
                // Simulate liveness detection process
                DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                    captureSession.stopRunning()
                    
                    // Simulate blink detection and 3D depth analysis
                    let isLive = self.simulateLivenessCheck()
                    completion(isLive, nil)
                }
            }
            
        } catch {
            completion(false, error)
        }
    }
    
    private func performFingerprintLivenessDetection(completion: @escaping (Bool, Error?) -> Void) {
        // Simulate pulse detection through Touch ID sensor
        // In real implementation, this would analyze blood flow patterns
        DispatchQueue.global().asyncAfter(deadline: .now() + 1.0) {
            let isLive = self.simulatePulseDetection()
            DispatchQueue.main.async {
                completion(isLive, nil)
            }
        }
    }
    
    private func performVoiceLivenessDetection(completion: @escaping (Bool, Error?) -> Void) {
        // Simulate voice challenge-response
        // In real implementation, this would generate random phrases and analyze speech patterns
        DispatchQueue.global().asyncAfter(deadline: .now() + 1.5) {
            let isLive = self.simulateVoiceChallenge()
            DispatchQueue.main.async {
                completion(isLive, nil)
            }
        }
    }
    
    private func performHeartRateLivenessDetection(completion: @escaping (Bool, Error?) -> Void) {
        // Simulate heart rate validation through camera
        // In real implementation, this would use camera and flash to detect pulse
        DispatchQueue.global().asyncAfter(deadline: .now() + 2.0) {
            let isLive = self.simulateHeartRateValidation()
            DispatchQueue.main.async {
                completion(isLive, nil)
            }
        }
    }
    
    // MARK: - Simulation Methods
    
    private func simulateLivenessCheck() -> Bool {
        // Simulate realistic liveness detection with 95% accuracy
        return Double.random(in: 0...1) < 0.95
    }
    
    private func simulatePulseDetection() -> Bool {
        // Simulate pulse detection with 90% accuracy
        return Double.random(in: 0...1) < 0.90
    }
    
    private func simulateVoiceChallenge() -> Bool {
        // Simulate voice challenge with 92% accuracy
        return Double.random(in: 0...1) < 0.92
    }
    
    private func simulateHeartRateValidation() -> Bool {
        // Simulate heart rate validation with 88% accuracy
        return Double.random(in: 0...1) < 0.88
    }
    
    // MARK: - Helper Methods
    
    func generateVoiceChallenge() -> String {
        let challenges = [
            "Please say: The quick brown fox jumps over the lazy dog",
            "Please say: Security is our top priority",
            "Please say: Today is a beautiful day for authentication",
            "Please say: My voice is my unique identifier",
            "Please say: Biometric security protects my data"
        ]
        return challenges.randomElement() ?? challenges[0]
    }
    
    func validateBlinkPattern(blinkTimings: [TimeInterval]) -> Bool {
        // Validate natural blink patterns
        // Normal blink rate: 15-20 blinks per minute
        // Blink duration: 100-400ms
        guard blinkTimings.count >= 2 else { return false }
        
        let intervals = zip(blinkTimings, blinkTimings.dropFirst()).map { $1 - $0 }
        let averageInterval = intervals.reduce(0, +) / Double(intervals.count)
        
        // Check if blink rate is within normal range (2-6 seconds between blinks)
        return averageInterval >= 2.0 && averageInterval <= 6.0
    }
    
    func analyzeFacialMovement(movementData: [(x: Float, y: Float)]) -> Bool {
        // Analyze natural facial movements
        guard movementData.count >= 5 else { return false }
        
        let movements = zip(movementData, movementData.dropFirst()).map { current, next in
            let dx = next.x - current.x
            let dy = next.y - current.y
            return sqrt(dx * dx + dy * dy)
        }
        
        let totalMovement = movements.reduce(0, +)
        let averageMovement = totalMovement / Float(movements.count)
        
        // Check for natural micro-movements (not too static, not too erratic)
        return totalMovement > 5.0 && averageMovement < 50.0
    }
}

// MARK: - AVCaptureVideoDataOutputSampleBufferDelegate
extension LivenessDetector: AVCaptureVideoDataOutputSampleBufferDelegate {
    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        // Process video frames for liveness detection
        // In real implementation, this would analyze frames for:
        // 1. Blink detection
        // 2. Facial movement analysis
        // 3. 3D depth estimation
        // 4. Texture analysis for spoofing detection
    }
}

enum LivenessError: Error {
    case cameraNotAvailable
    case detectionFailed
    case timeout
    
    var localizedDescription: String {
        switch self {
        case .cameraNotAvailable:
            return "Camera not available for liveness detection"
        case .detectionFailed:
            return "Liveness detection failed"
        case .timeout:
            return "Liveness detection timed out"
        }
    }
}
